Bugfix: Fixed the issue of the cancel button behavior in user security tab

AhsanSarwar0413 opened the pull request:
#### Summary

I have fixed the cancel button behavior in the security settings of user tab. I have fixed the behavior that when the user opens the password expandable section and click on the cancel button without doing any change in the passwords input, the current password input validation was triggered and user required to do two clicks on the cancel button in order to Unfold the password section which was wrong behavior and the validation of current password was triggered. I have fixed that issue, now when the user expand the password section and click on the cancel button for the first time it unfolds the password which is the right UI behavior. For more information PFA gifs.

#### Ticket Link
It is a small UI change with less than 20 lines. So, there is no need for ticket.

#### Screenshots
|  before  |  after  |
|----|----|

![before](https://github.com/user-attachments/assets/3dcf8629-d84d-41bf-8dc9-ba8f032ece35)
![after](https://github.com/user-attachments/assets/573ac73a-9edc-4b18-85c9-c7a366b5099c)



 

#### Release Note
NONE

Examples:
NONE


```release-note
NONE
```

mattermost-build commented:
Hello @AhsanSarwar0413,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

marianunez commented:
@AhsanSarwar0413 it seems that some tests are failing. Can you take a look?

yasserfaraazkhan commented:
/update-branch

AhsanSarwar0413 commented:
@marianunez @yasserfaraazkhan I have updated the snapshots in the codebase that causes the co pipeline failed. Please run the pipelines again. Thanks.

yasserfaraazkhan commented:
/e2e-tests

devinbinnie commented:
> hey @AhsanSarwar0413 thanks for identifying the issue and thanks for sending a solution. After analizing the problem and reproducing it myself I can tell is a little bit more complex than it initially looks and the solution might be too. Also, this same problem is happening for sure in other places, I have already identified the `profile modal > username|email` forms.
> 
> So, the technical explanation for the issue is the timing/order of the events and this happens because all the password inputs have onBlur event attached which trigger the validation and the onBlur event triggers before the onClick and nullifies it. My best guest is that the nullification is due to a react re-rendering that causes that the DOM element targeted for the event is recreated and this breaks the event chain.
> 
> That's why the onMouseDown serves as a workaround, because it fires before the onBlur. However, while this works, it's not semantically correct and could cause accessibility issues. I have been exploring other potential solutions today and haven't found an easy way to fix this, so my proposal is to go ahead with this workaround but also keep the onClick for keyboard access and touch devices, and open a ticket to fix this issue properly. Thoughts @devinbinnie

@pvev Since this is kind of a small fix and the bug doesn't tremendously break the app, I'm open to have the community member try an alternative solution. I'm hesitant to add a band-aid here as I feel like we'll be patching over something potentially causing other issues without understanding the root cause. So I don't think we should go with the `onMouseDown` fix.

If you can come up with a better fix feel free to share though :)

pvev commented:
> > hey @AhsanSarwar0413 thanks for identifying the issue and thanks for sending a solution. After analizing the problem and reproducing it myself I can tell is a little bit more complex than it initially looks and the solution might be too. Also, this same problem is happening for sure in other places, I have already identified the `profile modal > username|email` forms.
> > So, the technical explanation for the issue is the timing/order of the events and this happens because all the password inputs have onBlur event attached which trigger the validation and the onBlur event triggers before the onClick and nullifies it. My best guest is that the nullification is due to a react re-rendering that causes that the DOM element targeted for the event is recreated and this breaks the event chain.
> > That's why the onMouseDown serves as a workaround, because it fires before the onBlur. However, while this works, it's not semantically correct and could cause accessibility issues. I have been exploring other potential solutions today and haven't found an easy way to fix this, so my proposal is to go ahead with this workaround but also keep the onClick for keyboard access and touch devices, and open a ticket to fix this issue properly. Thoughts @devinbinnie
> 
> @pvev Since this is kind of a small fix and the bug doesn't tremendously break the app, I'm open to have the community member try an alternative solution. I'm hesitant to add a band-aid here as I feel like we'll be patching over something potentially causing other issues without understanding the root cause. So I don't think we should go with the `onMouseDown` fix.
> 
> If you can come up with a better fix feel free to share though :)

ok, well, my point was that even thought is not a big issue it is affecting just more than one input as the issue initially described and that is not just a trivial fix of stopping or preventing bubbling/propagation in someplace. So, I investigated a little bit more and followed the issue until the Input component and I think the proper solution would be to deferring the validation call to the next event loop cycle by surrounding the validate() https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/components/widgets/inputs/input/input.tsx#L133 call with a timeout to make sure the validation runs after the input loses focus and do not mess with the events timing. 

```
    const handleOnBlur = (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
        setFocused(false);
        setTimeout(() => {
            validateInput();
        }, 0); // Use setTimeout to make sure validation runs after the input loses focus and not mess with events timing

        if (onBlur) {
            onBlur(event);
        }
    };
```

Let me know what do you think @devinbinnie @AhsanSarwar0413

AhsanSarwar0413 commented:
@devinbinnie @pvev first of all apologies for the late response, I have also thought that I am optimistic regarding, with the help of mutation observer we can also achieve that. Please guys let me know after the discussion. So, I will fix it and update the PR.

pvev commented:
> I'm personally not a huge fan of setTimeout as a solution as it relies on arbitrary timings to work correctly

@devinbinnie me neither, I am not a fan of setTimeout when you have to rely in a "best guest" timing, but in this case is just for running the validation after the events have happened, running the validation in the next tick. So, for this situation using the setTimeout is an acceptable solution IMO. Also, since Input is such a wide used component the timeout is a very low intrusive solution that can be considered regressions-safe solution. But if @AhsanSarwar0413 can come up with a nice solution with the mutation observer, that would be even better.

AhsanSarwar0413 commented:
@pvev @devinbinnie I have updated the PR and I have maybe improvise this solution incase of input's `handleOnBlur` validation should not be activated if the cancel button is pressed. I have also reverted my old changes. Updated the following code because cancel button already having the specific id and targeting that id would make more sense. Let me know your thoughts guys?


```
const handleOnBlur = (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
        setFocused(false);
        const relatedTarget = event?.relatedTarget as HTMLElement;
        const isCancelAction = (relatedTarget && relatedTarget?.id === "cancelSetting") ?? false;
        if (!isCancelAction) {
            validateInput();
        }

        if (onBlur) {
            onBlur(event);
        }
    };

```

devinbinnie commented:
> @pvev @devinbinnie I have updated the PR and I have maybe improvise this solution incase of input's `handleOnBlur` validation should not be activated if the cancel button is pressed. I have also reverted my old changes. Updated the following code because cancel button already having the specific id and targeting that id would make more sense. Let me know your thoughts guys?
> 
> ```
> const handleOnBlur = (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
>         setFocused(false);
>         const relatedTarget = event?.relatedTarget as HTMLElement;
>         const isCancelAction = (relatedTarget && relatedTarget?.id === "cancelSetting") ?? false;
>         if (!isCancelAction) {
>             validateInput();
>         }
> 
>         if (onBlur) {
>             onBlur(event);
>         }
>     };
> ```

@AhsanSarwar0413 Thanks for updating the PR and working with us on this. Personally I would avoid focusing on a specific id here for a couple reasons:
1) We could have cases where we want the `Input` not to validate for a few different buttons. Since ids are supposed to be unique, we couldn't guarantee this would work on more than one element
2) It's more likely that other bugs would crop up as a result of requiring a specific named id, since it's hard for developers to discover that this is what we need to do. I believe this makes the fix a bit *too* narrow.

Not sure if @pvev has an opinion here, but I think going with the pattern I suggested above might be the way forward here, even if it's also not perfect.

AhsanSarwar0413 commented:
> setFocused(false);
> 
>         if (event.relatedTarget && (event.relatedTarget as HTMLElement).click) {
>             const target = event.relatedTarget as HTMLElement;
>             const listener = () => {
>                 validateInput();
>                 if (target) {
>                     target.removeEventListener('click', listener);
>                 }
>             };
>             target.addEventListener('click', listener);
>         } else {
>             validateInput();
>         }
> 
>         if (onBlur) {
>             onBlur(event);
>         }
>     };

@devinbinnie as far as I understand your code. How would you make sure that the onClick in relatedtarget is of `cancel` and not of `save` button. thats why I have done it specifically incase of cancel button. Because my solution is problem specific. Please correct me if I am wrong. Thanks.

AhsanSarwar0413 commented:
@pvev @devinbinnie any update? Do you guys wanted me to try out more solutions? Let me know?

AhsanSarwar0413 commented:
> > @pvev @devinbinnie any update? Do you guys wanted me to try out more solutions? Let me know?
> 
> hey @AhsanSarwar0413 did you already tried implementing the solution suggested by @devinbinnie ?

Yes, I have implemented and updated the PR.

devinbinnie reviewed (approved):
Thanks for keeping on this @AhsanSarwar0413 🚀 !

yasserfaraazkhan commented:
@AhsanSarwar0413 can you check the CI failures. Thank you.

AhsanSarwar0413 commented:
@yasserfaraazkhan updated the code after fixing ci issues.

AhsanSarwar0413 commented:
@yasserfaraazkhan fixed the remaining linter issues. Please run ci again and let me know if there is still any lint issue remain

AhsanSarwar0413 commented:
@yasserfaraazkhan Apologies, look like this one is overlooked. I have updated the PR and fixed the last issue that remains for ci-check-lint. You may run CI again.

yasserfaraazkhan commented:
/update-branch

yasserfaraazkhan commented:
@pvev @devinbinnie will be merging this PR.
